src/inmanta/validation_type.py

```diff
@@ -16,13 +16,58 @@
     Contact: code@inmanta.com
 """
 import importlib
+import inspect
 from collections import abc
-from typing import Optional
+from typing import Annotated, Optional, Type
 
 import pydantic
 
 from inmanta.stable_api import stable_api
-from inmanta.types import PrimitiveTypes
+from inmanta.types import PrimitiveTypes, PythonRegex
+
+
+@stable_api
+def parametrize_type(
+    base_type: Type[object] | abc.Callable[..., Type[object]],
+    validation_parameters: Optional[abc.Mapping[str, object]] = None,
+    *,
+    type_name: Optional[str] = None,
+) -> object:
+    """
+    Add validation parameters to the given type, if it supports any. Returns the parametrized type as a pydantic compatible
+    type, possibly through the use of typing.Annotated.
+
+    :param base_type: The type to parametrize. Either a type or a callable that produces a type when parametrized.
+    :param validation_parameters: Additional parameters to construct the validation type. Should only be passed if base_type
+        is a type generator function.
+    :param type_name: Optionally, the human-readable name of the type, for error messages.
+
+    :raises ValueError: When invalid parameters are passed.
+    :raises TypeError: When the given validation parameters are not valid with respect to the requested type.
+    """
+    type_name = type_name if type_name is not None else repr(base_type)
+
+    custom_annotations: list[object] = []
+    validation_parameters = dict(validation_parameters) if validation_parameters is not None else {}
```

Comment:
I'm curious, why cast to `dict` here ?

Reply:
To make a copy because we modify it later on